### PR TITLE
deliver SIGFPE to thread on divide-by-zero

### PR DESCRIPTION
**Pull request by jsultond**
This change was made to add compatibility for the SIGFPE signal,
specifically in the case of a Divide by Zero interrupt. It adds code
to the default handler to handle SIGFPE and updates the signal test
to also test for Divide by Zero.

### CODE REVIEW

File: test/runtime/signal.c [hunk 1]
```diff
@@ -591,6 +590,59 @@ test_sigsegv(void)
...
+
+    if (pthread_create(&pt, NULL, sigfpe_thread, (void *)0))
+      fail_perror("sigfpe_thread pthread_create");
+
+    sigtest_debug("yielding until child tid reported...\n");
+    yield_for(&child_tid);
+
+    if (pthread_join(pt, &retval))
+      fail_perror("blocking test pthread_join");
+  }
```
Comment (francescolavra) [inline test/runtime/signal.c:636]: No need to enclose these instructions in an inner {} block. Also, in this and other functions, please put the opening brace after the function parameter list on a new line, as done in the rest of the functions, and use a 4-space indentation style.

File: test/runtime/signal.c [hunk 2]
```diff
@@ -591,6 +590,59 @@ test_sigsegv(void)
...
+  if (arg == 0) {
+
+    /* generate sigfpe */
+    asm volatile("push   %rax \t\n\
+                    xor     %rax, %rax \t\n\
+                    idiv    %rax \t\n\
+                    ");
+  } else {
+    asm volatile("hlt");
+  }
```
Comment (francescolavra) [inline test/runtime/signal.c:611]: No need to check for `arg == 0`, and no need for the `hlt` instruction: just execute the sigfpe-generating instructions.

File: src/unix/unix.c [hunk 3]
```diff
@@ -114,27 +123,36 @@ define_closure_function(1, 1, context, default_fault_handler,
...
-                /* schedule this thread to either run signal handler or terminate */
-                schedule_frame(frame);
-                return 0;
-            } else {
-                rprintf("\nUnhandled page fault in kernel mode: ");
-                goto bug;
-            }
+    if (frame[FRAME_VECTOR] == 0) {
+      if (current_cpu()->state == cpu_user) {
+        deliver_fault_signal(SIGFPE, current, 0, FPE_INTDIV);
```
Comment (francescolavra) [inline src/unix/unix.c:128]: You should pass `fault_address(frame)` as `vaddr` argument to deliver_fault_signal().
Also, do not change the indentation of the the existing code.

**Review by francescolavra — COMMENTED** [review]
Thanks for the PR. Throughout (most of) the code we use a 4-space indentation style, could you use that as well (and not change the existing code if it already follows this style)?

File: src/unix/signal.c [hunk 4]
```diff
@@ -263,10 +263,10 @@ void deliver_signal_to_thread(thread t, struct siginfo *info)
 {
     int sig = info->si_signo;
     sig_debug("tid %d, sig %d\n", t->tid, sig);
-    if ((sig != SIGSEGV && sig != SIGKILL && sig != SIGSTOP) &&
+    if ((sig != SIGSEGV && sig != SIGKILL && sig != SIGSTOP && sig != SIGFPE) &&
         sig_is_ignored(t->p, sig)) {
-        sig_debug("signal ignored; no queue\n");
-        return;
+      sig_debug("signal ignored; no queue\n");
+      return;
```
[the remark this answers is not in the document]
Comment (jsultond) [reply src/unix/signal.c]: Hi. Thanks for reviewing my PR. The git-pre-commit-format script is actually what caused my code, and the little bits of the other code, to change from 4-space to 2-space.

File: src/unix/unix.c [hunk 5]
```diff
@@ -114,13 +126,22 @@ define_closure_function(1, 1, context, default_fault_handler,
...
-    if (frame[FRAME_VECTOR] == 14) {
-        u64 vaddr = fault_address(frame);
+    u64 vaddr = fault_address(frame);
+    if (frame[FRAME_VECTOR] == 0) {
+        if (current_cpu()->state == cpu_user) {
+            deliver_fault_signal(SIGFPE, current, vaddr, FPE_INTDIV);
+            schedule_frame(frame);
+            return 0;
+        } else {
+            rprintf("\nSIGFPE occurs in kernel mode\n");
```
[the remark this answers is not in the document]
Comment (jsultond) [reply src/unix/unix.c]: I've changed it to print out the Divide by zero error rather than SIGFPE.

File: src/unix/unix.c [hunk 6]
```diff
@@ -114,13 +126,22 @@ define_closure_function(1, 1, context, default_fault_handler,
...
+            goto bug;
+        }
+    } else if (frame[FRAME_VECTOR] == 14) {
         vmap vm = vmap_from_vaddr(p, vaddr);
         if (vm == INVALID_ADDRESS) {
             if (user) {
-                pf_debug("no vmap found for addr 0x%lx, rip 0x%lx", vaddr, frame[FRAME_RIP]);
+                pf_debug("no vmap found for addr 0x%lx, rip 0x%lx", vaddr,
```
[the remark this answers is not in the document]
Comment (jsultond) [reply src/unix/unix.c]: I've unwrapped this one and a similar one below it.
